pandas/io/formats/info.py

```diff
@@ -209,147 +275,353 @@ def info(self) -> None:
+        """Select verbose table builder: with or without non-null counts."""
+        if self.show_counts:
+            return DataFrameTableBuilderVerboseWithCounts
+        else:
+            return DataFrameTableBuilderVerboseNoCounts
 
-        if self.memory_usage:
-            # append memory usage of df to display
-            size_qualifier = ""
-            if self.memory_usage == "deep":
-                deep = True
-            else:
-                # size_qualifier is just a best effort; not guaranteed to catch
-                # all cases (e.g., it misses categorical data even with object
-                # categories)
-                deep = False
-                if "object" in counts or self.data.index._is_memory_usage_qualified():
-                    size_qualifier = "+"
-            mem_usage = self._get_mem_usage(deep=deep)
-            lines.append(f"memory usage: {_sizeof_fmt(mem_usage, size_qualifier)}\n")
-        fmt.buffer_put_lines(self.buf, lines)
 
+class TableBuilderAbstract(ABC):
+    """Abstract builder for info table.
 
-class DataFrameInfo(BaseInfo):
-    def _get_mem_usage(self, deep: bool) -> int:
-        return self.data.memory_usage(index=True, deep=deep).sum()
+    Parameters
+    ----------
+    info : BaseInfo
+        Instance of DataFrameInfo or SeriesInfo.
+    """
 
-    def _get_ids_and_dtypes(self) -> Tuple["Index", "Series"]:
-        return self.data.columns, self.data.dtypes
+    _lines: List[str]
 
-    def _verbose_repr(
-        self, lines: List[str], ids: "Index", dtypes: "Series", show_counts: bool
-    ) -> None:
-        col_count = len(ids)
-        lines.append(f"Data columns (total {col_count} columns):")
-
-        id_head = " # "
-        column_head = "Column"
-        col_space = 2
-
-        max_col = max(len(pprint_thing(k)) for k in ids)
-        len_column = len(pprint_thing(column_head))
-        space = max(max_col, len_column) + col_space
-
-        max_id = len(pprint_thing(col_count))
-        len_id = len(pprint_thing(id_head))
-        space_num = max(max_id, len_id) + col_space
-
-        if show_counts:
-            counts = self.data.count()
-            if col_count != len(counts):  # pragma: no cover
-                raise AssertionError(
-                    f"Columns must equal counts ({col_count} != {len(counts)})"
-                )
-            count_header = "Non-Null Count"
-            len_count = len(count_header)
-            non_null = " non-null"
-            max_count = max(len(pprint_thing(k)) for k in counts) + len(non_null)
-            space_count = max(len_count, max_count) + col_space
-            count_temp = "{count}" + non_null
+    def __init__(self, *, info):
+        self.info = info
+
+    @abstractmethod
+    def get_lines(self) -> List[str]:
+        """Product in a form of list of lines (strings)."""
+
+
+class DataFrameTableBuilder(TableBuilderAbstract):
```

Comment:
can you reduce the builder abstraction a bit. this is too java-esque. you can just pass parameterize / use partial functions. I find this style too different from the rest of pandas. let's make this simpler.

Reply:
As you requested, I dropped ``DataFrameTableBuilderVerboseNoCounts`` and ``DataFrameTableBuilderVerboseWithCounts``.
Instead, I parametrized ``DataFrameTableBuilderVerbose`` (with_counts = True/False). I personally prefer the original implementation, but this one is also OK. My concerns with the new implementation are rather minor. Two if statements checking for ``with_counts`` in two separate functions (fragile), and presence of ``_gen_non_null_counts``, which is required for the table with counts and is irrelevant for table without counts.

I leave ``TableBuilderAbstract`` as @MarcoGorelli is working on series info, thus it will be reasonable for the builders to have the same interface.

Reply:
Regarding Series.info - if you want to work on it I'm happy for you to take it over (else I'll go back to it once this PR is in)

Reply:
@MarcoGorelli, I can work on Series.info from the code point of view. Since you already have documentation and tests, it seems to be quite manageable for me.

Reply:
Cool, I'll close that PR then! If for whatever reason you change your mind, please ping me and I'll take it up again

Reply:
Good. I will start working on series info once this is PR merged.